### PR TITLE
Minor improvements for Target<T>

### PR DESCRIPTION
**Pull request by drieseng**
* Minor performance improvement for `Target<T>` constructor.
* Reduces allocations in `Target<T>.ReadConstraintFromTarget()` when more than one **ConstraintAttribute** is applied to a member.

**Before:**

|                                               Method |        Mean |      Error |     StdDev | Gen 0/1k Op | Gen 1/1k Op | Gen 2/1k Op | Allocated Memory/Op |
|----------------------------------------------------- |------------:|-----------:|-----------:|------------:|------------:|------------:|--------------------:|
|                    Constructor |    63.50 ns |  1.2426 ns |  1.1015 ns |      0.0801 |           - |           - |               336 B |
|  Constraint_MoreThanOneConstraintAttributes_AllMatch | 2,234.96 ns | 14.2570 ns | 13.3360 ns |      0.2213 |           - |           - |               944 B |

**After:**

|                                              Method |        Mean |     Error |    StdDev | Gen 0/1k Op | Gen 1/1k Op | Gen 2/1k Op | Allocated Memory/Op |
|---------------------------------------------------- |------------:|----------:|----------:|------------:|------------:|------------:|--------------------:|
|                   Constructor|    48.18 ns | 0.1601 ns | 0.1337 ns |      0.0801 |           - |           - |               336 B |
| Constraint_MoreThanOneConstraintAttributes_AllMatch | 2,115.73 ns | 5.5302 ns | 5.1730 ns |      0.1945 |           - |           - |               824 B |

### CODE REVIEW

File: src/Ninject/Planning/Targets/Target.cs [hunk 1]
```diff
@@ -55,8 +55,8 @@ protected Target(MemberInfo member, T site)
             this.Member = member;
             this.Site = site;
 
-            this.constraint = new Lazy<Func<IBindingMetadata, bool>>(this.ReadConstraintFromTarget);
-            this.isOptional = new Lazy<bool>(this.ReadOptionalFromTarget);
+            this.constraint = new Lazy<Func<IBindingMetadata, bool>>(() => this.ReadConstraintFromTarget());
```
Comment (scott-xu) [inline src/Ninject/Planning/Targets/Target.cs]: @drieseng Can you please help me understand why this change could improve the performance?

Comment (drieseng) [reply src/Ninject/Planning/Targets/Target.cs]: This is related to the use of method group vs. delegate.
A delegate is cached by the compiler, method group is not.
I'll update this PR (for Array.Empty<T>()) and re-run the benchmarks again to make sure the results are "stable".

Comment (scott-xu) [reply src/Ninject/Planning/Targets/Target.cs]: Thanks!

Comment (drieseng) [reply src/Ninject/Planning/Targets/Target.cs]: I verified that the (minor) performance improvement for the constructor is indeed from replacing a method group by a delegate.

**Comment by drieseng** [comment]
@scott-xu Can we get this in?
